### PR TITLE
Checking for disk corruption

### PR DESCRIPTION
**Pull request by zed-rishabh**
Checking for disk corruption by comparing the output from qemu-img and
format passed in config.

Signed-off-by: zed-rishabh <rgupta@zededa.com>

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 1]
```diff
@@ -1821,6 +1837,11 @@ func configToXencfg(config types.DomainConfig, status types.DomainStatus,
 		if ds.Format == zconfig.Format_CONTAINER {
 			continue
 		}
+		err := checkDiskFormat(ds)
+		if err != nil {
+			log.Errorf("%v", err)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go]: FWIW you can just go a log.Error(err) here.

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 2]
```diff
@@ -2266,17 +2287,22 @@ func DomainCreate(status types.DomainStatus) (int, string, error) {
 	filename := xenCfgFilename(status.AppNum)
 	mpFileName := mountPointFileName(status.AppNum)
 	log.Infof("DomainCreate %s ... xenCfgFilename - %s\n", status.DomainName, filename)
+	var containerImageSha256 string
+	for _, ds := range status.DiskStatusList {
+		if ds.Format == zconfig.Format_CONTAINER {
+			containerImageSha256 = ds.ImageSha256
+			continue
```
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go]: In the old code you set containerImageSha256 based on the first container, and with this change you are setting it to the last container. Does that matter? Intentional?

Comment (zed-rishabh) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: It doesn't matter because of manifest will either one or none container disks.

**Comment by zed-rishabh** [comment]
> Does this check detect the cases we've had when the qemu image was corrupted? In that case do we get some additional errors from qemu-img which we can report, it is it just telling us that it isn't a qcow2?

Currently, I am not aware of the errors returned by the qemu in case of corruption. In these changes, We are checking the disk format provided in the config and output from the qemu.

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (zed-rishabh) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: For me, it is failing with the following error:
# github.com/lf-edge/eve/pkg/pillar/cmd/domainmgr
cmd/domainmgr/domainmgr.go:1842:14: cannot use err (type error) as type string in argument to logrus.Errorf

**Comment by kalyan-nidumolu** [comment]
> > Does this check detect the cases we've had when the qemu image was corrupted? In that case do we get some additional errors from qemu-img which we can report, it is it just telling us that it isn't a qcow2?
> 
> Currently, I am not aware of the errors returned by the qemu in case of corruption. In these changes, we are checking only the disk format provided in the config and output from the qemu. We are checking format at the time of creating xen configuration and domain creation.

Hi Rishabh, We have seen the Qemu Errors after a reboot. So here is what happened:

1) For VMs, we reboot - There were some changes in flight when the reboot happened
      - We did not wait for those changes to get flushed to the disk and this corrupted the RW copy
          of the VM disk

2) After reboot, when we start the VM, Qemu complains that the Disk is no longer a valid Qcow2 disk

The issue is similar to what you are trying to fix for Volumes. The question is - does this code path detect the error for VMs also? Or should we add that check separately?

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: log.Error(err) not log.Errorf
We use that pattern in lots of places.

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: OK

**Comment by rvs** [comment]
Hey @zed-rishabh -- I think without the check action on qemu-disk this is only a partial fix. How about you add that before we merge this?

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 1 above)
Comment (zed-rishabh) [reply pkg/pillar/cmd/domainmgr/domainmgr.go]: Oh, my bad. I will update it.

**Comment by zed-rishabh** [comment]
> > > Does this check detect the cases we've had when the qemu image was corrupted? In that case do we get some additional errors from qemu-img which we can report, it is it just telling us that it isn't a qcow2?
> > 
> > 
> > Currently, I am not aware of the errors returned by the qemu in case of corruption. In these changes, we are checking only the disk format provided in the config and output from the qemu. We are checking format at the time of creating xen configuration and domain creation.
> 
> Hi Rishabh, We have seen the Qemu Errors after a reboot. So here is what happened:
> 
> 1. For VMs, we reboot - There were some changes in flight when the reboot happened
>    
>    * We did not wait for those changes to get flushed to the disk and this corrupted the RW copy
>      of the VM disk
> 2. After reboot, when we start the VM, Qemu complains that the Disk is no longer a valid Qcow2 disk
> 
> The issue is similar to what you are trying to fix for Volumes. The question is - does this code path detect the error for VMs also? Or should we add that check separately?

Hi Kalyan, this check will work for both containers and VM at the time of domain creation.

**Comment by zed-rishabh** [comment]
> Hey @zed-rishabh -- I think without the check action on qemu-disk this is only a partial fix. How about you add that before we merge this?

Sure @rvs, I will explore and add check action on qemu-disk.

**Comment by zed-rishabh** [comment]
> Lets talk about qemu check

Sure.

**Comment by rvs** [comment]
This looks like it is reasonably ready to be merged. What do you think @nordmark @kalyan-nidumolu ?

**Pull request merged**